File: src/services/source/providers/ScribbleHubProvider.ts
import { load } from "cheerio";
import { sanitizeTitle } from "../../../utils/stringUtils";
import { fetchFormPage } from "../../network/fetcher";
import { ChapterInfo, NovelMetadata, SourceProvider } from "../types";

const SCRIBBLE_HUB_BASE_URL = "https://www.scribblehub.com";
const SCRIBBLE_HUB_AJAX_URL =
  "https://www.scribblehub.com/wp-admin/admin-ajax.php";
const SCRIBBLE_HUB_MAX_TOC_PAGES = 500;

const toAbsoluteUrl = (href: string, baseUrl: string): string => {
  if (!href) return "";

  try {
    return new URL(href, baseUrl).toString();
  } catch {
    return href;
  }
};

const unique = (values: string[]): string[] => {
  return Array.from(new Set(values.filter(Boolean)));
};

const stableFallbackId = (url: string): string => {
  return `sh_url_${encodeURIComponent(url.toLowerCase())}`;
};

const normalizeAjaxPageHtml = (html: string): string => {
  return html.replace(/0\s*$/, "").trim();
};

const parseTocItems = (html: string, baseUrl: string): ChapterInfo[] => {
  const $ = load(html);
  const chapters: ChapterInfo[] = [];

  $(".toc_ol li").each((_, element) => {
    const link = $(element).find('a[href*="/read/"][href*="/chapter/"]').first();
    const href = link.attr("href");
    if (!href) return;

    const url = toAbsoluteUrl(href, baseUrl);
    const title = sanitizeTitle(link.text().trim());
    const chapterId = ScribbleHubProvider.getChapterId?.(url);

    chapters.push({
      id: chapterId,
      title: title || "Untitled Chapter",
      url,
    });
  });

  return chapters;
};

const parsePaginationPageCount = (html: string): number => {
  const $ = load(html);
  const pageNumbers: number[] = [];

  $("#pagination-mesh-toc li").each((_, element) => {
    const pageNumber = Number.parseInt($(element).text().trim(), 10);
    if (!Number.isNaN(pageNumber)) {
      pageNumbers.push(pageNumber);
    }
  });

  return pageNumbers.length > 0 ? Math.max(...pageNumbers) : 1;
};

export const ScribbleHubProvider: SourceProvider = {
  name: "Scribble Hub",
  baseUrl: SCRIBBLE_HUB_BASE_URL,

  isSource: (url: string) => {
    return /https?:\/\/(?:www\.)?scribblehub\.com\/series\/\d+/i.test(url);
  },

  getStoryId: (url: string): string => {
    const match = url.match(/\/series\/(\d+)/i);
    if (match) {
      return `sh_${match[1]}`;
    }
    return stableFallbackId(url);
  },

  getChapterId: (url: string): string | undefined => {
    const match = url.match(/\/chapter\/(\d+)/i);
    return match ? `sh_${match[1]}` : undefined;
  },

  parseMetadata: (html: string): NovelMetadata => {
    const $ = load(html);

    const title =
      $(".fic_title").first().text().trim() ||
      $(".wi_fic_title").first().text().trim() ||
      $("h1").first().text().trim() ||
      $('meta[property="og:title"]').attr("content") ||
      "Unknown Title";

    const author =
      $(".auth_name_fic a").first().text().trim() ||
      $('a[href*="/profile/"]').first().text().trim() ||
      $('meta[name="author"]').attr("content") ||
      "Unknown Author";

    const coverUrl =
      $(".fic_image img").first().attr("src") ||
      $(".fic_image img").first().attr("data-src") ||
      $('meta[property="og:image"]').attr("content") ||
      undefined;

    const description =
      $(".wi_fic_desc").first().text().trim() ||
      $(".fic_synopsis").first().text().trim() ||
      $("#synopsis").first().text().trim() ||
      $('meta[property="og:description"]').attr("content") ||
      $('meta[name="description"]').attr("content") ||
      undefined;

    const canonicalUrl =
      $('link[rel="canonical"]').attr("href") ||
      $('meta[property="og:url"]').attr("content") ||
      undefined;

    const tags: string[] = [];
    $(
      '.wi_fic_genre a, .wi_fic_tags a, .fic_genre a, .fic_tags a, a[href*="/genre/"], a[href*="/tag/"]',
    ).each((_, element) => {
      const tag = $(element).text().trim();
      if (tag) tags.push(tag);
    });
    const uniqueTags = unique(tags);

    const score =
      $(".numscore").first().text().trim() ||
      $('[itemprop="ratingValue"]').first().text().trim() ||
      $(".rate_fic_user").first().text().trim() ||
      undefined;

    return {
      title,
      author,
      coverUrl,
      description,
      tags: uniqueTags.length > 0 ? uniqueTags : undefined,
      score,
      canonicalUrl,
    };
  },

  getChapterList: async (
    html: string,
    baseUrl: string,
    onProgress?: (message: string) => void,
  ): Promise<ChapterInfo[]> => {
    onProgress?.("Parsing chapter list...");

    const $ = load(html);
    const postId = $("#mypostid").attr("value");
    const totalPages = parsePaginationPageCount(html);
    const chapters = parseTocItems(html, baseUrl);
    const seenChapterUrls = new Set(chapters.map((chapter) => chapter.url));
    const shouldFetchPaginatedChapters =
      Boolean(postId) && (totalPages > 1 || chapters.length >= 15);

    if (shouldFetchPaginatedChapters && postId) {
      for (let page = 2; page <= SCRIBBLE_HUB_MAX_TOC_PAGES; page++) {
        const progressLabel =
          totalPages > 1
            ? `Fetching chapter page ${page}/${totalPages}${
                page > totalPages ? "+" : ""
              }...`
            : `Fetching chapter page ${page}...`;

        onProgress?.(progressLabel);

        const pageHtml = await fetchFormPage(SCRIBBLE_HUB_AJAX_URL, {
          action: "wi_getreleases_pagination",
          pagenum: page,
          mypostid: postId,
        });

        const pageChapters = parseTocItems(
          normalizeAjaxPageHtml(pageHtml),
          baseUrl,
        );

        if (pageChapters.length === 0) {
          break;
        }

        const newChapters = pageChapters.filter((chapter) => {
          if (seenChapterUrls.has(chapter.url)) {
            return false;
          }

          seenChapterUrls.add(chapter.url);
          return true;
        });

        if (newChapters.length === 0) {
          break;
        }

        chapters.push(...newChapters);

        if (pageChapters.length < 15) {
          break;
        }
      }
    }

    return [...chapters].reverse();
  },

  parseChapterContent: (html: string): string => {
    const $ = load(html);
    const content = $("#chp_raw").first();

    content.find("script").remove();
    content.find("style").remove();
    content.find(".wi_authornotes").remove();
    content.find(".sharedaddy").remove();
    content.find(".code-block").remove();

    return content.html()?.trim() || "No content found";
  },
};

File: android/app/src/main/java/com/vinicius741/webnovelarchiver/source/ScribbleHubProvider.kt
package com.vinicius741.webnovelarchiver.source

import com.vinicius741.webnovelarchiver.domain.archive.PercentEncoding
import com.vinicius741.webnovelarchiver.domain.model.ChapterInfo
import com.vinicius741.webnovelarchiver.domain.model.NovelMetadata
import com.vinicius741.webnovelarchiver.domain.model.PublicationStatus
import com.vinicius741.webnovelarchiver.domain.model.SourceMetadata
import com.vinicius741.webnovelarchiver.domain.model.SourceMetric
import com.vinicius741.webnovelarchiver.domain.model.SourceMetricKind
import com.vinicius741.webnovelarchiver.source.network.NetworkClient
import com.vinicius741.webnovelarchiver.source.network.NetworkParseException
import com.vinicius741.webnovelarchiver.source.network.SourceAccessBlockedException
import com.vinicius741.webnovelarchiver.source.network.SourceChapterListIncompleteException
import com.vinicius741.webnovelarchiver.source.network.SourceNetworkPolicy
import kotlinx.coroutines.CancellationException
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import timber.log.Timber

@Suppress("TooManyFunctions")
object ScribbleHubProvider : SourceProvider {
    override val descriptor =
        SourceDescriptor(
            id = "scribble_hub",
            displayName = "Scribble Hub",
            browseUrl = "https://www.scribblehub.com",
            hosts = setOf("scribblehub.com"),
            capabilities = SourceCapabilities(latestChapterSync = true),
            networkPolicy =
                SourceNetworkPolicy(
                    minimumRequestGapMillis = 3_000L,
                    maximumAttempts = 3,
                    retryableStatusCodes = setOf(403, 429),
                    maximumRequestsPerWindow = 12,
                ),
            managesBrowserSession = true,
            cookieSeeds =
                listOf(
                    SourceCookieSeed(
                        url = "https://www.scribblehub.com",
                        value = "toc_show=50; Domain=.scribblehub.com; Path=/; Max-Age=31536000",
                    ),
                ),
            renderedPageRules =
                listOf(
                    SourceRenderedPageRule(pathContains = "/chapter/", requiredSelector = "#chp_raw"),
                    SourceRenderedPageRule(
                        pathContains = "/series/",
                        requiredSelector = "#mypostid, .fic_title, .wi_fic_title",
                    ),
                    SourceRenderedPageRule(pathSuffix = "/wp-admin/admin-ajax.php", allowEmptyBody = true),
                ),
            featuredMetrics =
                listOf(
                    SourceMetricKind.READERS,
                    SourceMetricKind.TOTAL_VIEWS,
                    SourceMetricKind.WORDS,
                ),
        )
    private const val AJAX_URL = "https://www.scribblehub.com/wp-admin/admin-ajax.php"

    override fun classifyUrl(url: String): SourceUrlKind? {
        val path = sourcePath(url) ?: return null
        return when {
            Regex("""^/series/\d+(?:/[^/]+)?/?$""", RegexOption.IGNORE_CASE).matches(path) -> SourceUrlKind.STORY
            Regex("""^/read/[^/]+/chapter/\d+(?:/[^/]+)?/?$""", RegexOption.IGNORE_CASE).matches(path) ->
                SourceUrlKind.CHAPTER
            else -> null
        }
    }

    override fun getStoryId(url: String) =
        Regex("/series/(\\d+)", RegexOption.IGNORE_CASE)
            .find(url)
            ?.groupValues
            ?.get(1)
            ?.let { "sh_$it" }
            ?: "sh_url_${PercentEncoding.encodeURIComponent(url.lowercase())}"

    override fun getChapterId(url: String) =
        Regex("/chapter/(\\d+)", RegexOption.IGNORE_CASE)
            .find(url)
            ?.groupValues
            ?.get(1)
            ?.let { "sh_$it" }

    override fun parseMetadata(html: String): NovelMetadata {
        val doc = Jsoup.parse(html, baseUrl)
        val title =
            firstText(doc, ".fic_title", ".wi_fic_title", "h1").ifBlank {
                doc.selectFirst("meta[property=og:title]")?.attr("content")
                    ?: "Unknown Title"
            }
        val author =
            firstText(doc, ".auth_name_fic a", ".auth_name_fic", "a[href*=/profile/]").ifBlank {
                doc.selectFirst("meta[name=author]")?.attr("content")
                    ?: "Unknown Author"
            }
        val cover =
            (doc.selectFirst(".fic_image img")?.absUrl("src") ?: "")
                .ifBlank { doc.selectFirst(".fic_image img")?.absUrl("data-src").orEmpty() }
                .ifBlank { doc.selectFirst("meta[property=og:image]")?.attr("content").orEmpty() }
                .ifBlank { null }
        val description =
            (
                firstStructuredText(doc, ".wi_fic_desc", ".fic_synopsis", "#synopsis")
                    .ifBlank {
                        doc.selectFirst("meta[property=og:description]")?.attr("content")
                            ?: doc.selectFirst("meta[name=description]")?.attr("content").orEmpty()
                    }.stripScribbleHubSynopsisToggle()
            ).ifBlank { null }
        val canonical = doc.selectFirst("link[rel=canonical]")?.absUrl("href") ?: doc.selectFirst("meta[property=og:url]")?.attr("content")
        val tags =
            doc
                .select(".wi_fic_genre a, .wi_fic_tags a, .fic_genre a, .fic_tags a, a[href*=/genre/], a[href*=/tag/]")
                .map {
                    it.text().trim()
                }.filter { it.isNotBlank() }
                .distinct()
                .toMutableList()
        val score = scribbleHubScore(doc)
        val patreonUrl = findPatreonUrl(doc)
        val publicationStatus = scribbleHubPublicationStatus(doc)
        val contentWarnings =
            doc
                .select("li.mature_contains > a[href*=/content/]")
                .map { normalizedSourceText(it.text()) }
                .filter { it.isNotBlank() }
                .distinct()
                .toMutableList()
        return NovelMetadata(
            title = title,
            author = author,
            coverUrl = cover,
            description = description,
            tags = tags.ifEmpty { null },
            score = score,
            canonicalUrl = canonical,
            patreonUrl = patreonUrl,
            publicationStatus = publicationStatus,
            sourceMetadata =
                SourceMetadata(
                    publishedAt = scribbleHubParsePublicationDate(doc),
                    contentWarnings = contentWarnings,
                ),
        )
    }

    override suspend fun enrichMetadata(
        metadata: NovelMetadata,
        html: String,
        url: String,
        network: NetworkClient,
        progress: (String) -> Unit,
    ): NovelMetadata {
        val statsUrl = scribbleHubCanonicalSeriesUrl(metadata.canonicalUrl ?: url)?.let { "$it/stats/" } ?: return metadata
        return try {
            progress("Fetching Scribble Hub statistics...")
            val stats = parseStatsMetadata(network.fetch(statsUrl))
            if (stats.metrics.isEmpty()) {
                metadata
            } else {
                metadata.copy(sourceMetadata = mergeScribbleHubSourceMetadata(metadata.sourceMetadata, stats))
            }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            // Statistics are supplemental. A blocked, missing, or changed stats page must not make
            // the base series import fail, but log so regressions in stats parsing don't vanish.
            Timber.w(error, "Scribble Hub stats enrichment failed for %s; continuing without metrics.", statsUrl)
            metadata
        }
    }

    /** Parses the public `/stats/` page without relying on its current positional layout. */
    internal fun parseStatsMetadata(html: String): SourceMetadata {
        val doc = Jsoup.parse(html, baseUrl)
        val labeledValues = scribbleHubExtractLabeledValues(doc)
        val metrics =
            scribbleHubStatLabels
                .values
                .distinct()
                .mapNotNull { kind ->
                    labeledValues[kind]?.let { value ->
                        parseSourceMetricValue(value)?.let { SourceMetric(kind = kind, value = it) }
                    }
                }.toMutableList()
        val ratings = scribbleHubParseRatingsCount(doc)
        if (ratings != null && metrics.none { it.kind == SourceMetricKind.RATINGS }) {
            metrics += SourceMetric(kind = SourceMetricKind.RATINGS, value = ratings)
        }
        return SourceMetadata(metrics = metrics)
    }

    override suspend fun getChapterList(
        html: String,
        url: String,
        network: NetworkClient,
        progress: (String) -> Unit,
    ): List<ChapterInfo> {
        progress("Parsing chapter list...")
        val doc = Jsoup.parse(html, url)
        val postId = doc.selectFirst("#mypostid")?.attr("value")
        val inline = parseToc(doc)
        if (postId.isNullOrBlank() || inline.size < 15) {
            progress("${chapterCountLabel(inline.size)} found")
            return inline.asReversed()
        }
        progress("Fetching chapter page 1 · ${chapterCountLabel(inline.size)} found...")
        // The paginated list is the authoritative full list: a blocked or truncated page must
        // fail the fetch instead of returning a partial list that a full sync would treat as
        // "author removed chapters" (R03). Latest-only paths keep their fallback behavior.
        val firstPage =
            try {
                fetchTocPage(network, url, postId, 1)
            } catch (error: SourceAccessBlockedException) {
                throw incompleteTocException(1, error)
            }
        val start = if (firstPage.size >= inline.size) firstPage else inline
        val chapters = paginateTocPages(start, { page -> fetchTocPage(network, url, postId, page) }, progress)
        progress("${chapterCountLabel(chapters.size)} found")
        return chapters.asReversed()
    }

    override suspend fun getLatestChapterList(
        html: String,
        url: String,
        network: NetworkClient,
        progress: (String) -> Unit,
    ): List<ChapterInfo>? {
        val doc = Jsoup.parse(html, url)
        val postId = doc.selectFirst("#mypostid")?.attr("value")
        val chapters = parseToc(doc)
        val latest =
            if (!postId.isNullOrBlank() && chapters.size >= 15) {
                progress("Fetching latest chapter page...")
                try {
                    fetchTocPage(network, url, postId, 1).ifEmpty { chapters }
                } catch (_: SourceAccessBlockedException) {
                    chapters
                }
            } else {
                chapters
            }
        return latest.asReversed()
    }

    override fun parseChapterContent(html: String): String {
        val doc = Jsoup.parse(html)
        // See RoyalRoadProvider.parseChapterContent: throw instead of returning a placeholder so the
        // failure surfaces as a download-job error rather than getting embedded in the EPUB.
        val content = doc.selectFirst("#chp_raw") ?: throw NetworkParseException("Chapter content not found on page")
        content.select("script, style, .wi_authornotes, .sharedaddy, .code-block").remove()
        return content.html().trim()
    }

    private fun parseToc(doc: Document): List<ChapterInfo> =
        doc.select(".toc_ol li").mapNotNull { li ->
            val link = li.selectFirst("a[href*=/read/][href*=/chapter/]") ?: return@mapNotNull null
            val href = link.absUrl("href").ifBlank { link.attr("href") }
            ChapterInfo(
                getChapterId(href),
                sanitizeTitle(link.text()).ifBlank {
                    "Untitled Chapter"
                },
                href,
                publishedAt = li.chapterPublishedAt(),
            )
        }

    private fun chapterCountLabel(count: Int): String = "$count ${if (count == 1) "chapter" else "chapters"}"

    private suspend fun fetchTocPage(
        network: NetworkClient,
        url: String,
        postId: String,
        page: Int,
    ): List<ChapterInfo> {
        // toc_show=50 is seeded into CookieManager at app startup and carried by the shared
        // AndroidCookieJar; no manual Cookie header here (it would be overwritten by OkHttp's
        // BridgeInterceptor once a cf_clearance cookie is also in the jar).
        val pageHtml =
            network
                .postForm(
                    AJAX_URL,
                    mapOf(
                        "action" to "wi_getreleases_pagination",
                        "pagenum" to page,
                        "mypostid" to postId,
                    ),
                ).replace(Regex("0\\s*$"), "")
                .trim()
        return parseToc(Jsoup.parse(pageHtml, url))
    }

    private fun firstText(
        doc: Document,
        vararg selectors: String,
    ): String =
        selectors
            .firstNotNullOfOrNull { selector ->
                doc.selectFirst(selector)?.text()?.trim()?.takeIf {
                    it.isNotBlank()
                }
            }.orEmpty()

    // Like [firstText], but preserves the original paragraph/line-break layout of the matched
    // element via [blockText]. Used for long free form fields (e.g. synopsis) where collapsing
    // `<p>`/`<br>` into a single space-joined line would discard the author's structure.
    private fun firstStructuredText(
        doc: Document,
        vararg selectors: String,
    ): String =
        selectors
            .firstNotNullOfOrNull { selector ->
                doc.selectFirst(selector)?.blockText()?.takeIf { it.isNotBlank() }
            }.orEmpty()

    private fun scribbleHubScore(doc: Document): String? =
        doc
            .select("script[type=application/ld+json]")
            .asSequence()
            .map { it.data().ifBlank { it.html() } }
            .filter { it.contains("AggregateRating") }
            .mapNotNull {
                Regex(""""ratingValue"\s*:\s*"?([^",}]+)"?""")
                    .find(it)
                    ?.groupValues
                    ?.get(1)
                    ?.trim()
            }.firstOrNull { it.isNotBlank() }
            ?: firstText(doc, ".numscore", "[itemprop=ratingValue]", ".rate_fic_user").ifBlank { null }
            ?: doc
                .selectFirst("#ratefic_user")
                ?.text()
                ?.let { Regex("""\d+(?:\.\d+)?(?=\s*\()""").find(it)?.value }

    private fun scribbleHubPublicationStatus(doc: Document): PublicationStatus {
        val explicit =
            doc
                .select(
                    ".wi_fic_status, .fic_status, .fic_status_info, .fic_status_label, " +
                        ".series-status, .story-status, .post-status",
                ).map { it.text().trim() }
                .firstNotNullOfOrNull(::publicationStatusFromSourceText)
        if (explicit != null) return explicit

        val bodyText = doc.body()?.text().orEmpty()
        val updatedLineStatus =
            Regex("""(?i)\b(Completed|Complete|Ongoing|Hiatus|Hiatused|Dropped|Cancelled|Canceled)\b\s*[-–—]\s*Updated\b""")
                .find(bodyText)
                ?.groupValues
                ?.getOrNull(1)
        val rightsStatusPattern =
            Regex(
                """(?i)\bAll\s+Rights\s+Reserved\b[\s;·•|,.-]{0,40}""" +
                    """(Completed|Complete|Ongoing|Hiatus|Hiatused|Dropped|Cancelled|Canceled)\b""",
            )
        val rightsSidebarStatus =
            rightsStatusPattern
                .find(bodyText)
                ?.groupValues
                ?.getOrNull(1)
        return publicationStatusFromSourceText(updatedLineStatus)
            ?: publicationStatusFromSourceText(rightsSidebarStatus)
            ?: PublicationStatus.unknown
    }
}
